core/base/mergeTreePrincipalGeodesicsDecoding/MergeTreePrincipalGeodesicsDecoding.h

```diff
@@ -0,0 +1,484 @@
+      bool isSecondInput = false) {
+      auto &vSToUse = (isSecondInput ? pTrees2Vs_ : pVS_);
+      auto &v2sToUse = (isSecondInput ? pTrees2V2s_ : pV2s_);
+      auto vSizeToUse = (isSecondInput ? vSize2_ : vSize_);
+
+      // Preprocessing
+      preprocessBarycenter<dataType>(barycenter);
+
+      std::array<double, 2> middle;
+      getGeodesicsMiddle<dataType>(
+        barycenter, vSToUse, v2sToUse, vSizeToUse, middle);
+
+      // Construct geodesics trees
+      geodesicsTrees.resize(
+        vSToUse.size(), std::vector<ftm::MergeTree<dataType>>(k_));
+      tGeodesics_.resize(
+        geodesicsTrees.size(),
+        std::vector<std::vector<double>>(
+          geodesicsTrees[0].size(), std::vector<double>(vSToUse.size(), 0.0)));
+#ifdef TTK_ENABLE_OPENMP
+#pragma omp parallel for schedule(dynamic) num_threads(this->threadNumber_)
+#endif
+      for(unsigned int i = 0; i < geodesicsTrees.size(); ++i)
+        for(unsigned int j = 0; j < geodesicsTrees[i].size(); ++j) {
+          double t = 1.0 / (k_ - 1) * j;
+          getInterpolation<dataType>(barycenter, vSToUse[i], v2sToUse[i],
+                                     vSizeToUse, t, geodesicsTrees[i][j]);
+          tGeodesics_[i][j][i] = t;
+          int i2 = (i + 1) % 2;
+          tGeodesics_[i][j][i2] = middle[i2];
+        }
+
+      // Postprocessing
+      postprocessingPipeline<dataType>(&(barycenter.tree));
+#ifdef TTK_ENABLE_OPENMP
+#pragma omp parallel for schedule(dynamic) num_threads(this->threadNumber_)
+#endif
+      for(unsigned int i = 0; i < geodesicsTrees.size(); ++i)
+        for(unsigned int j = 0; j < geodesicsTrees[i].size(); ++j)
+          postprocessingPipeline<dataType>(&(geodesicsTrees[i][j].tree));
+    }
+
+    template <class dataType>
+    void constructGeodesicsEllipses(
+      ftm::MergeTree<dataType> &barycenter,
+      std::vector<ftm::MergeTree<dataType>> &geodesicsEllipses,
+      bool isSecondInput = false) {
+      auto &vSToUse = (isSecondInput ? pTrees2Vs_ : pVS_);
+      auto &v2sToUse = (isSecondInput ? pTrees2V2s_ : pV2s_);
+      auto vSizeToUse = (isSecondInput ? vSize2_ : vSize_);
+
+      // Preprocessing
+      preprocessBarycenter<dataType>(barycenter);
+
+      // Init
+      unsigned int noSample = k_ * 2;
+      geodesicsEllipses.resize(noSample);
+      tEllipses_.resize(geodesicsEllipses.size());
+
+      std::vector<std::vector<double *>> vS(2), v2s(2);
+      vS[0] = vSToUse[0];
+      vS[1] = vSToUse[1];
+      v2s[0] = v2sToUse[0];
+      v2s[1] = v2sToUse[1];
+
+      // Get middle of geodesic
+      std::array<double, 2> middle;
+      getGeodesicsMiddle<dataType>(barycenter, vS, v2s, vSizeToUse, middle);
+
+      // Get ellipses
+#ifdef TTK_ENABLE_OPENMP
+#pragma omp parallel for schedule(dynamic) num_threads(this->threadNumber_)
+#endif
+      for(unsigned int i = 0; i < noSample; ++i) {
+        double angle = 360.0 / noSample * i;
+        double pi = 3.14159265359;
```

Comment:
please use the system constant `M_PI` instead.

Reply:
Done in the next commit!